### PR TITLE
[SLP][REVEC] Make ShuffleCostEstimator and ShuffleInstructionBuilder can vectorize vector instructions.

### PR DESCRIPTION
**Pull request by HanKuanChen**
Since the mask indices expect the source is scalar type, we need to transform the mask indices into a form which can be used when REVEC is enabled. The transform is only called before the CreateShuffleVector.

This is a following patch for https://github.com/llvm/llvm-project/pull/99499.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -253,6 +253,21 @@ static FixedVectorType *getWidenedType(Type *ScalarTy, unsigned VF) {
...
 }
 
+static void transformScalarShuffleIndiciesToVector(unsigned VecTyNumElements,
+                                                   SmallVectorImpl<int> &Mask) {
+  // The ShuffleBuilder implementation use shufflevector to splat an "element".
+  // But the element have different meaning for SLP (scalar) and REVEC
+  // (vector). We need to expand Mask into masks which shufflevector can use
+  // directly.
+  SmallVector<int> NewMask(Mask.size() * VecTyNumElements);
+  for (size_t I = 0, E = Mask.size(); I != E; ++I)
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:263]:
```suggestion
  for (unsigned I : seq<unsigned>(Mask.size()))
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -253,6 +253,21 @@ static FixedVectorType *getWidenedType(Type *ScalarTy, unsigned VF) {
...
+    for (unsigned J = 0; J != VecTyNumElements; ++J)
+      NewMask[I * VecTyNumElements + J] = Mask[I] == PoisonMaskElem
+                                              ? PoisonMaskElem
+                                              : Mask[I] * VecTyNumElements + J;
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:267]:
```suggestion
    for (auto [I, MaskV] : enumerate(MutableArrayRef(NewMask).slice(I * VecTyNumElements, VecTyNumElements)))
      MaskV = Mask[I] == PoisonMaskElem
                                              ? PoisonMaskElem
                                              : Mask[I] * VecTyNumElements + J;

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -8800,13 +8815,17 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
       // Shuffle single vector.
       ExtraCost += GetValueMinBWAffectedCost(V1);
       CommonVF = cast<FixedVectorType>(V1->getType())->getNumElements();
+      if (auto *VecTy = dyn_cast<FixedVectorType>(ScalarTy))
+        CommonVF /= VecTy->getNumElements();
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8819]: Add assertion that CommonVF is large enough

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 4]
```diff
@@ -8800,13 +8815,17 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
       assert(
           all_of(Mask,
                  [=](int Idx) { return Idx < static_cast<int>(CommonVF); }) &&
           "All elements in mask must be less than CommonVF.");
     } else if (V1 && !V2) {
       // Shuffle vector and tree node.
       unsigned VF = cast<FixedVectorType>(V1->getType())->getNumElements();
+      if (auto *VecTy = dyn_cast<FixedVectorType>(ScalarTy))
+        VF /= VecTy->getNumElements();
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8828]: assertion

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 5]
```diff
@@ -8833,6 +8852,8 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
     } else if (!V1 && V2) {
       // Shuffle vector and tree node.
       unsigned VF = cast<FixedVectorType>(V2->getType())->getNumElements();
+      if (auto *VecTy = dyn_cast<FixedVectorType>(ScalarTy))
+        VF /= VecTy->getNumElements();
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8856]: assertion. Better to have a function, doing this, to avoid duplication

**Review by alexey-bataev — COMMENTED** [review]
Need the tests

**Comment by HanKuanChen** [comment]
> Need the tests

This PR and https://github.com/llvm/llvm-project/pull/99499 are trying to solve the same test (test3 in `llvm/test/Transforms/SLPVectorizer/revec.ll`).
Do I need to merge this into https://github.com/llvm/llvm-project/pull/99499?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 3 above)
Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: If `CommonVF` is not large enough, the assert should fail.
```
      assert(all_of(Mask,
                    [=](int Idx) {
                      return Idx < 2 * static_cast<int>(CommonVF);
                    }) &&
             "All elements in mask must be less than 2 * CommonVF.");
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 6]
```diff
@@ -8728,6 +8744,10 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
       }
       return TTI::TCC_Free;
     };
+    auto getVF = [&](Value *V) {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8747]:
```suggestion
    auto GetVF = [&](Value *V) {
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 3 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Anyway better to have assertion here

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 7]
```diff
@@ -8728,6 +8744,16 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
       }
       return TTI::TCC_Free;
     };
+    auto GetVF = [&](Value *V) {
+      unsigned VNumElements =
+          cast<FixedVectorType>(V->getType())->getNumElements();
+      assert(VNumElements > ScalarTyNumElements &&
+             VNumElements % ScalarTyNumElements == 0);
+      // V1 and V2 would be the argument of shufflevector.
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8752]: Add assertion message

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 8]
```diff
@@ -8728,6 +8744,16 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+      // The maximum index of Mask should be (VNumElements * 2) - 1.
+      assert(VNumElements <= 2 * ScalarTyNumElements * Mask.size());
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8754]: Add assertion message

**Review by alexey-bataev — COMMENTED** [review]
Some previous comments not addressed yet

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 3 above)
Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Add assertion in `GetVF`.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 4 above)
Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Add assertion in `GetVF`.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 5 above)
Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Add assertion in `GetVF`.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 3 above)
Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Add assertion in `GetVF`.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 9]
```diff
@@ -8728,6 +8744,15 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
       }
       return TTI::TCC_Free;
     };
+    auto GetVF = [&](Value *V) {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8747]: Better make it a function and use here and in other places

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 10]
```diff
@@ -8728,6 +8744,15 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+      unsigned VNumElements =
+          cast<FixedVectorType>(V->getType())->getNumElements();
+      assert(VNumElements > ScalarTyNumElements &&
+             "the number of elements of V is not large enough");
+      assert(VNumElements % ScalarTyNumElements == 0 &&
```
[the remark this answers is not in the document]
Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8752]: `ScalarTyNumElements` can be non-power-of-2. But the rule `VNumElements % ScalarTyNumElements` is still applied.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 9 above)
Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Add a `getVF` member function. (see https://github.com/llvm/llvm-project/pull/99606/commits/c40998e669c009b32147d66c8d7af18be23fd2e4)

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 11]
```diff
@@ -8271,6 +8286,27 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
   /// to move most of the long shuffles cost estimation to TTI.
   bool SameNodesEstimated = true;
 
+  /// V is expected to be a vectorized value.
+  /// When REVEC is disabled, there is no difference between VF and
+  /// VNumElements.
+  /// When REVEC is enabled, VF is VNumElements / ScalarTyNumElements.
+  /// e.g., if ScalarTy is <4 x Ty> and V1 is <8 x Ty>, 2 is returned instead
+  /// of 8.
+  unsigned getVF(Value *V) {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8295]:
```suggestion
  unsigned getVF(Value *V) const {
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 12]
```diff
@@ -8271,6 +8286,27 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+    assert(V != nullptr && "V cannot be nullptr");
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8296]:
```suggestion
    assert(V && "V cannot be nullptr");
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 13]
```diff
@@ -8271,6 +8286,27 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+    assert(isa<FixedVectorType>(V->getType()) &&
+           "V does not have FixedVectorType");
+    assert(ScalarTy != nullptr && "ScalarTy cannot be nullptr");
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8299]:
```suggestion
    assert(ScalarTy && "ScalarTy cannot be nullptr");
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 14]
```diff
@@ -12179,6 +12217,8 @@ class BoUpSLP::ShuffleInstructionBuilder final : public BaseShuffleAnalysis {
         }
     }
     int VF = cast<FixedVectorType>(V1->getType())->getNumElements();
+    if (auto *VecTy = dyn_cast<FixedVectorType>(ScalarTy))
+      VF /= VecTy->getNumElements();
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:12221]: Should this be processed by something similar like getVF or getVF itself?

**Review by alexey-bataev — COMMENTED** [review]
> > Need the tests
> 
> This PR and #99499 are trying to solve the same test (test3 in `llvm/test/Transforms/SLPVectorizer/revec.ll`). Do I need to merge this into #99499?

Probably, the patch is not acceptable without tests

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 14 above)
Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: We can make `getVF` into a protected member function of `BaseShuffleAnalysis`. However, `BaseShuffleAnalysis` would need to store `ScalarTy`. See https://github.com/llvm/llvm-project/pull/99606/commits/f1de612dbf027f53e1f7dff1616aa1a98f2acd26.

**Comment by HanKuanChen** [comment]
> > > Need the tests
> > 
> > 
> > This PR and #99499 are trying to solve the same test (test3 in `llvm/test/Transforms/SLPVectorizer/revec.ll`). Do I need to merge this into #99499?
> 
> Probably, the patch is not acceptable without tests

Will merge into #99499 once this PR is good.

**Comment by alexey-bataev** [comment]
The changes looks good but need the test

**Comment by HanKuanChen** [comment]
Close the PR and move into https://github.com/llvm/llvm-project/pull/99499.

**Pull request closed without merging**